Learner checkpoint : Frontend implementation of saved progress

paprajapati9 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR implements the frontend changes for storing checkpoint progress of logged in user. TDD : [Link](https://docs.google.com/document/d/1eWHs46cOHcm7NuQl8pM9Dada9nTgg3GoCQLgQJSNmIY/)
2. This PR does the following: Stores and fetches the checkpoints progress of the user and redirects the user to the most recently reached checkpoint on initial load of exploration.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes 
made in this PR work correctly. If this PR is for a developer-facing feature, 
provide the videos/screenshots of developer-facing interface. 

Please also include videos/screenshots of the developer tools 
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof 
for the changes then please leave a comment "No proof of changes needed because {{Reason}}" 
and remove all the sections below.
-->

https://user-images.githubusercontent.com/42764756/166180159-cd64efeb-9974-40c0-a26f-0bb5fb8ffa15.mp4


Previously completed state card view in the next sessions

https://user-images.githubusercontent.com/42764756/166659833-3ff1f3ba-cd4e-494c-a703-e2734494211b.mp4




#### Proof of changes in Arabic language

<!--
If the PR changes the UI in any of the files listed in rtl_css.py (i.e, those that have 
a separate .rtl.css file for styling), make sure to add screenshots with the site 
language set to Arabic as well (we use Arabic as it is a language written from right to left).  
-->
![image](https://user-images.githubusercontent.com/42764756/166179637-702bc407-fc4c-4588-92ea-ebc0dde82f04.png)




## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

Manan-Rathi committed:
Started

Manan-Rathi committed:
Add data

Manan-Rathi committed:
Add progress-bar

Manan-Rathi committed:
Add divisions to progess-bar.

Manan-Rathi committed:
Added lesson authors dropdown.

Manan-Rathi committed:
cleaning

Manan-Rathi committed:
Added files to frontend denylist

Manan-Rathi committed:
Fix conflicts and issues

Manan-Rathi committed:
Modified existing files and fix bugs.

Manan-Rathi committed:
Merge branch 'develop' of https://github.com/oppia/oppia into lesson-info

Manan-Rathi committed:
Added spec for lesson-info-component.

Manan-Rathi committed:
Fixed lint & Modyfied userJourneys.js.

Manan-Rathi committed:
Lint fix-2

Manan-Rathi committed:
Modyfied spec for exp-footer.

Manan-Rathi committed:
minor fix

Manan-Rathi committed:
Css improvements & remove unused classes from exp-footer.

Manan-Rathi committed:
Made requested changes.

Manan-Rathi committed:
lint fix

Manan-Rathi committed:
fix rtl error

Manan-Rathi committed:
Merge branch 'develop' of https://github.com/oppia/oppia into lesson-info

Manan-Rathi committed:
type errors

Manan-Rathi committed:
Made dropdown list scrollable

Manan-Rathi committed:
lint fix

Manan-Rathi committed:
changes

Manan-Rathi committed:
fix

Manan-Rathi committed:
changes

Manan-Rathi committed:
Merge branch 'develop' of https://github.com/oppia/oppia into lesson-info

Manan-Rathi committed:
lint-fix

Manan-Rathi committed:
minor fix

Manan-Rathi committed:
Merge branch 'develop' of https://github.com/oppia/oppia into lesson-info

Manan-Rathi committed:
Changes-2

Manan-Rathi committed:
Initial commit.

Manan-Rathi committed:
Record checkpoints.

Manan-Rathi committed:
Resolve ts-checks.

Manan-Rathi committed:
Added new feats.

Manan-Rathi committed:
Imrpoved feats.

Manan-Rathi committed:
Fix lints

Manan-Rathi committed:
Fix lints-2

Manan-Rathi committed:
Fix lint.

Manan-Rathi committed:
Added to frontend_denylist.

Manan-Rathi committed:
lint fix

Manan-Rathi committed:
fix conflicts

Manan-Rathi committed:
Altered field names and removed support for last_completed.

Manan-Rathi committed:
Added new feats.

Manan-Rathi committed:
Merge branch 'develop' of https://github.com/oppia/oppia into li-2

Manan-Rathi committed:
ts-fixes

Manan-Rathi committed:
nit

Manan-Rathi committed:
removed file

Manan-Rathi committed:
nit

paprajapati9 committed:
checking learner checkpoint

paprajapati9 committed:
added redirection to most recently checkpoint on load

paprajapati9 committed:
fixed answered state card

paprajapati9 committed:
fixed messsage

paprajapati9 committed:
pulled changes from existing PR

paprajapati9 committed:
Merge branch 'oppia:develop' into learner-checkpoint

paprajapati9 committed:
fixed rtl test

paprajapati9 committed:
lint fixes

paprajapati9 committed:
added tests to multiple files

IamEzio commented:
@paprajapati9 could you please also add the code for making the 'restart button' disabled until the full content of the first state loads. This will prevent user from sending a 'restart' request to the backend before a POST request is sent to the backend that creates exploration_user_data and marks the first state as the `most_recently_reached_checkpoint`. More context can be found [here](https://github.com/oppia/oppia/pull/15213#discussion_r848885118).
Thanks!

paprajapati9 committed:
removed saving checkpoint progress for iframes and disabled restart button for 0% progress

paprajapati9 commented:
> @paprajapati9 could you please also add the code for making the 'restart button' disabled until the full content of the first state loads. This will prevent user from sending a 'restart' request to the backend before a POST request is sent to the backend that creates exploration_user_data and marks the first state as the `most_recently_reached_checkpoint`. More context can be found [here](https://github.com/oppia/oppia/pull/15213#discussion_r848885118). Thanks!

Have disabled the restart button until at least 1 checkpoint is completed, this should also fix the scenario you have mentioned.

paprajapati9 committed:
fixed checkpoints progress and added rtl support for progress modal

paprajapati9 committed:
fixed leasson modal behaviour for logged in and out users

paprajapati9 committed:
resolved merge conflicts

paprajapati9 committed:
extended frontend coverage

paprajapati9 committed:
lint fixes

paprajapati9 committed:
extended frontend coverage

paprajapati9 committed:
fixed karma test

paprajapati9 committed:
Merge branch 'oppia:develop' into learner-checkpoint

nithusha21 reviewed (changes requested):
Hi @paprajapati9 could you do the following:
- Have one of the project leads of the project do a first pass
- Share the design doc/mocks so that we have something to review against (I have no context of this change for example)
- Are there UI changes? If so please get a UI review from the designers.

paprajapati9 committed:
fixes for learner checkpoints progress

paprajapati9 committed:
lint fixes

paprajapati9 committed:
frontend coverage

paprajapati9 commented:
@seanlip Please take a look. Thanks!

paprajapati9 committed:
minor change

seanlip commented on core/domain/user_services.py:
```diff
@@ -2313,14 +2313,23 @@ def _get_checkpoints_in_order(init_state_name, states):
         if current_state_name not in visited_state_names:
             visited_state_names.append(current_state_name)
             current_state = states[current_state_name]
+            # We treat the last state card also as a checkpoint to store
```
Actually I think it was intentional to omit this (I had thought about it originally). What's the point of restoring to the last state? There would be nothing else to play.

paprajapati9 replied:
Updated, we will not treat the last state as a checkpoint because as soon as the exploration is completed we clear the learner's progress and restart the exploration.

seanlip reviewed (changes requested):
Thanks! Reviewed 1 codeowner file.

seanlip reviewed (changes requested):
Thanks @paprajapati9! I think this is a good start.

On the code side, I only have one comment (inline).

On the UI side:
- There's a "flash of unstyled content" just before the full modal is shown (this happens several times in the video). Would it be possible to fix this? It is quite awkward.
- The modal styling is a bit off. I suggest that you take a screenshot of it and start a thread with Diana and me for corrections.
- Please drop the exclamation marks at the end of phrases. Also these texts should be internationalized.
- When the user restarts from a checkpoint, there's a flash of the initial card. That should be removed.
- When the user restarts from a checkpoint, there should probably be a toast/message on the top saying something like: "Continuing this lesson from where you left off". Please check with Diana for the exact text.
- Thanks for providing the RTL screenshot! The only question I have for that one is, shouldn't "Progress" be on the right-hand side and the bar be flipped?

Thanks!

kevintab95 commented on core/templates/pages/exploration-player-page/templates/lesson-information-card-modal.component.ts:
```diff
@@ -116,6 +122,15 @@ export class LessonInformationCardModalComponent extends ConfirmOrCancelModal {
+      this.explorationId
+    ).then(() => {
+      // Required for the put operation to deliver data to backend.
+      this.windowRef.nativeWindow.location.reload();
```
Is it possible to achieve the same result (i.e. restarting an exp), without a full refresh of the page?

paprajapati9 replied:
This could be done I think although it will be a bit tricky, but I am not sure if we want that kind of a behaviour. We want to give user a feeling of clearing progress and restarting the whole exploration altogether and I think reloading seems good from that perspective, but would love to hear your thoughts. Thanks!

kevintab95 replied:
I think from the user perspective, it should be the same if you programatically restart the exp (instead of doing a full refresh of the page.) In fact, it will be a bit faster and will benefit users having low speed internet connectivity.

paprajapati9 replied:
Ok @kevintab95, in that case let's do it in a seperate PR, let's first get this functionality up, it will take some time to implement it and we need to get this PR merged in quickly to atleast get the checkpoints functionality up and then we can do the revisions in a seperate PR. I have created an issue for this https://github.com/oppia/oppia/issues/15419

kevintab95 commented on core/templates/pages/exploration-player-page/services/exploration-engine.service.ts:
```diff
@@ -398,8 +405,8 @@ export class ExplorationEngineService {
 
-  getAuthorRecommendedExpIds(): string[] {
-    return this.exploration.getAuthorRecommendedExpIds(this.currentStateName);
+  getAuthorRecommendedExpIds(stateName: string): string[] {
```
Should we update the method name to reflect that it returns the recommended exp ids for the provided stateName (not the "current state")?

paprajapati9 replied:
Yes, have updated it's name. Thanks!

kevintab95 commented on core/templates/pages/exploration-player-page/services/exploration-engine.service.ts:
```diff
@@ -559,6 +568,100 @@ export class ExplorationEngineService {
   }
+
+  getStateCardByName(stateName: string): StateCard {
+    let _nextFocusLabel = this.focusManagerService.generateFocusLabel();
```
Prefer `const` if the variable is not reassigned.

paprajapati9 replied:
Done

kevintab95 commented on core/templates/pages/exploration-player-page/services/exploration-engine.service.ts:
```diff
@@ -559,6 +568,100 @@ export class ExplorationEngineService {
+    let contentHtml = (
+      this.exploration.getState(stateName).content.html
+    );
+    contentHtml = contentHtml + this._getRandomSuffix();
```
Can we inline contentHtml instead of reassigning the variable?

paprajapati9 replied:
Done

kevintab95 commented on core/templates/pages/exploration-player-page/services/exploration-engine.service.ts:
```diff
@@ -559,6 +568,100 @@ export class ExplorationEngineService {
+    contentHtml = contentHtml + this._getRandomSuffix();
+    interactionHtml = interactionHtml + this._getRandomSuffix();
+
+    let stateCard = StateCard.createNewCard(
```
Prefer `const` or inline it.

paprajapati9 replied:
done

kevintab95 commented on core/templates/pages/exploration-player-page/services/exploration-engine.service.ts:
```diff
@@ -559,6 +568,100 @@ export class ExplorationEngineService {
+    let stateGraphLinks: { source: string; target: string }[] = [];
+
+    // Create a list of all possible links between states.
+    for (var stateName of Object.keys(allStates)) {
```
Prefer `let`

paprajapati9 replied:
done

paprajapati9 committed:
review changes

paprajapati9 commented:
@seanlip Please take a look. Thanks!

seanlip reviewed (changes requested):
Thanks @paprajapati9! No concerns for my (backend) codeowner files.

Remaining UI concerns:
- Modal styling is a bit off (let's discuss on email thread).
- The toast message for when the user restarts from a checkpoint is on the bottom-right ... and I completely missed it the first time while viewing the video, I think students will miss it too. Can it go in the top center?

Thanks!

kevintab95 commented on core/templates/pages/exploration-player-page/services/exploration-engine.service.ts:
```diff
@@ -559,6 +568,100 @@ export class ExplorationEngineService {
+    let shortestPathToState: string[] = [];
+    let queue: string[] = [];
+    let seen: Record<string, boolean> = {};
+    let parent: Record<string, string> = {};
```
Can we rename these so that it is more apparent what the variables represent?

paprajapati9 replied:
updated, thanks!

kevintab95 commented on core/templates/pages/exploration-player-page/services/exploration-engine.service.ts:
```diff
@@ -559,6 +568,100 @@ export class ExplorationEngineService {
+    }
+    // Reverse the path to go initStateName to destStateName.
+    shortestPathToState = shortestPathToState.reverse();
+    return shortestPathToState;
```
Perhaps just inline this since it isn't used later. It gets a bit confusing if you reassign the same variable but it represents a different thing.

paprajapati9 replied:
done

kevintab95 commented on core/templates/pages/exploration-player-page/learner-experience/tutor-card.component.ts:
```diff
@@ -277,6 +278,15 @@ export class TutorCardComponent {
+      this.displayedCard.getInteraction() &&
+      (this.prevSessionStatesProgress.indexOf(
+        this.displayedCard.getStateName()) !== -1
+      )
```
Seems like this could be passed in as a boolean input since all of this information is present in the parent component. Should we do that instead?

paprajapati9 replied:
yeah, made it like that. Thanks!

kevintab95 commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.component.ts:
```diff
@@ -764,9 +891,19 @@ export class ConversationSkinComponent {
       this.nextCard.getStateName());
+
+    // We do not store checkpoints progress for iframes hence we do not
+    // need to consider redirecting in that case.
```
Could you expand on or explain "redirecting" here please? It is not immediately clear what is being referred to.

paprajapati9 replied:
done

kevintab95 commented on core/templates/pages/exploration-player-page/layout-directives/exploration-footer.component.ts:
```diff
@@ -60,6 +62,9 @@ export class ExplorationFooterComponent {
   expInfo: LearnerExplorationSummaryBackendDict;
   completedWidth: number = 0;
   expStates: StateObjectsBackendDict;
+  completedCheckpoints: number = 0;
```
Since this is a number, maybe `completedCheckpointsCount` might be better.

paprajapati9 replied:
updated

kevintab95 commented on core/templates/pages/exploration-player-page/layout-directives/exploration-footer.component.ts:
```diff
@@ -60,6 +62,9 @@ export class ExplorationFooterComponent {
   completedWidth: number = 0;
   expStates: StateObjectsBackendDict;
+  completedCheckpoints: number = 0;
+  isLastCheckpointReached: boolean = false;
```
```suggestion
  lastCheckpointWasReached: boolean = false;
```

paprajapati9 replied:
updated

kevintab95 reviewed (changes requested):
Thanks @paprajapati9! I've left some comments but I think it looks great overall. PTAL

paprajapati9 committed:
review changes

paprajapati9 committed:
Merge branch 'oppia:develop' into learner-checkpoint

paprajapati9 committed:
minor fixes

paprajapati9 committed:
Merge branch 'learner-checkpoint' of https://github.com/paprajapati9/oppia into learner-checkpoint

seanlip commented:
@paprajapati9 just a heads up, most CI tests are failing. Might be worth looking into!

paprajapati9 committed:
Merge branch 'oppia:develop' into learner-checkpoint

paprajapati9 committed:
fix frontend karma tests

paprajapati9 committed:
lesson info modal changes

paprajapati9 committed:
Merge branch 'learner-checkpoint' of https://github.com/paprajapati9/oppia into learner-checkpoint

paprajapati9 committed:
lint fixes

paprajapati9 committed:
fix tests

paprajapati9 commented:
@kevintab95 Please take a look. Thanks!

paprajapati9 committed:
lesson info modal changes

paprajapati9 committed:
Merge branch 'learner-checkpoint' of https://github.com/paprajapati9/oppia into learner-checkpoint

paprajapati9 committed:
fix failing tests

paprajapati9 committed:
added lesson info modal tooltip

paprajapati9 commented:
@seanlip @kevintab95 @nithusha21 @aks681 The PR is complete. Can you guys please review it asap. Thanks!

seanlip commented on assets/i18n/en.json:
```diff
@@ -292,6 +292,8 @@
     "I18N_EXPLORATION_RvopsvVdIb0J_TITLE": "Tying Ratios to Actual Numbers",
     "I18N_EXPLORATION_SR1IKIdLxnm1_DESCRIPTION": "Ava got bored from playing the amusement park games, so Kamal created a fun math game. Can you beat Kamal's game? Click on this lesson to find out!",
     "I18N_EXPLORATION_SR1IKIdLxnm1_TITLE": "Recap: Variables",
+    "I18N_EXPLORATION_STARTING_FROM_CHECKPOINT": "Welcome back! Continuing this lesson from where you left off",
```
Missing period at end.

paprajapati9 replied:
Done

seanlip commented on assets/i18n/en.json:
```diff
@@ -292,6 +292,8 @@
     "I18N_EXPLORATION_SR1IKIdLxnm1_DESCRIPTION": "Ava got bored from playing the amusement park games, so Kamal created a fun math game. Can you beat Kamal's game? Click on this lesson to find out!",
     "I18N_EXPLORATION_SR1IKIdLxnm1_TITLE": "Recap: Variables",
+    "I18N_EXPLORATION_STARTING_FROM_CHECKPOINT": "Welcome back! Continuing this lesson from where you left off",
+    "I18N_EXPLORATION_STATE_PREVIOUSLY_COMPLETED": "You have already covered this in a previous session.",
```
Can you show this in the UI video proofs?

paprajapati9 replied:
Done

seanlip commented on assets/i18n/en.json:
```diff
@@ -567,6 +569,7 @@
     "I18N_LEARN_TOPIC": "Learn <[topicName]>",
     "I18N_LESSON_AUTHORS_DROPDOWN_TITLE": "Lesson Authors",
     "I18N_LESSON_INFO_HEADER": "Lesson Info",
+    "I18N_LESSON_INFO_TOOLTIP_MESSAGE": "Great job on reaching your second checkpoint! View your progress and other lesson information here.",
```
You've reached a checkpoint. Great job! View your progress and other lesson information here.

paprajapati9 replied:
Done

seanlip reviewed (changes requested):
Thanks @paprajapati9, this looks great! I just had minor comments on my codeowner files. 

Also, note the failing CI checks.

paprajapati9 committed:
review changes and e2e test test-fix

paprajapati9 committed:
Merge branch 'learner-checkpoint' of https://github.com/paprajapati9/oppia into learner-checkpoint

paprajapati9 committed:
lint fixes

paprajapati9 commented:
Hey @seanlip, have addressed your codeowner reviews, the left e2e tests are taking some time, I am looking into them currently. Thanks!

paprajapati9 committed:
test fix

paprajapati9 committed:
Merge branch 'oppia:develop' into learner-checkpoint

aks681 reviewed (approved):
Lgtm for one codeowner file

paprajapati9 committed:
test fixes

paprajapati9 committed:
Merge branch 'oppia:develop' into learner-checkpoint

paprajapati9 committed:
Merge branch 'learner-checkpoint' of https://github.com/paprajapati9/oppia into learner-checkpoint

paprajapati9 committed:
topic e2e test fix

paprajapati9 committed:
e2e tests fix

paprajapati9 committed:
minor fix

paprajapati9 committed:
e2e test changes

paprajapati9 committed:
e2e test changes

paprajapati9 committed:
minor changes

paprajapati9 committed:
minor change in tests

paprajapati9 committed:
minor change in tests

paprajapati9 committed:
lint fixes

paprajapati9 committed:
removed tooltip feature

paprajapati9 committed:
minor updates in lesson info modal

paprajapati9 commented:
@seanlip @nithusha21 @srijanreddy98 Please take a look. Thanks!

paprajapati9 commented:
Hey @seanlip @srijanreddy98 @nithusha21, can u guys please review this asap. The PR needs to be merged quickly. Thanks!

seanlip reviewed (approved):
Thanks! LGTM from my end.

Though @dchen97 should review too (for UI/UX). Adding her to this PR.

dchen97 commented:
Thanks @paprajapati9! Just a couple of questions as I didn't see them in the video:

1. What does it look like when a user tries to go back and see previously submitted answers?
2. What does it look like on the topic or classroom pages when a lesson is in progress? 
3. Have we implemented the [discoverability tool tip](https://docs.google.com/document/d/1VUQ7LvAIL8DwBbckUpFCZQJwDSbVEFBBxc2Xt9YWViQ/edit?pli=1#bookmark=id.gt2rdc39umj7) described in the design doc? 

Overall, I approve getting this merged for now but think we should as quickly as possible follow up with a separate PR addressing the following:

1. We need a message letting the user know that they're not about to enter the lesson from the beginning. We should also have some message letting them know they'll start at the beginning if they've already finished the lesson in a previous session, as it feels really jarring to keep the existing behavior without any explanation to the user.
2. From talking with Sean, it seems as though the existing lesson info modal is not very discoverable and we may want to change where / how the progress bar is presented. Until we get finalized UX mocks, though, I think we should make sure the existing modal has a clear explanation for what restarting a lesson means and it shouldn't be the prominent CTA in the modal.
3. We should combine the lesson info modals in the top and bottom parts of the lesson player. UX has proposed mocks already for this and will share shortly.
4. We should remove the dialog to ask if they're sure they want to reload the page.

paprajapati9 commented:
> Thanks @paprajapati9! Just a couple of questions as I didn't see them in the video:
> 
> 1. What does it look like when a user tries to go back and see previously submitted answers?
> 2. What does it look like on the topic or classroom pages when a lesson is in progress?
> 3. Have we implemented the [discoverability tool tip](https://docs.google.com/document/d/1VUQ7LvAIL8DwBbckUpFCZQJwDSbVEFBBxc2Xt9YWViQ/edit?pli=1#bookmark=id.gt2rdc39umj7) described in the design doc?
> 
> Overall, I approve getting this merged for now but think we should as quickly as possible follow up with a separate PR addressing the following:
> 
> 1. We need a message letting the user know that they're not about to enter the lesson from the beginning. We should also have some message letting them know they'll start at the beginning if they've already finished the lesson in a previous session, as it feels really jarring to keep the existing behavior without any explanation to the user.
> 2. From talking with Sean, it seems as though the existing lesson info modal is not very discoverable and we may want to change where / how the progress bar is presented. Until we get finalized UX mocks, though, I think we should make sure the existing modal has a clear explanation for what restarting a lesson means and it shouldn't be the prominent CTA in the modal.
> 3. We should combine the lesson info modals in the top and bottom parts of the lesson player. UX has proposed mocks already for this and will share shortly.
> 4. We should remove the dialog to ask if they're sure they want to reload the page.

@dchen97 Thanks for the reviews!
Regarding the things u mentioned:
1. The video for the previously submitted answers is there. The 2nd video in the proof section shows this.
2. We haven't really thought about how it looks like on the topic or classroom pages when a lesson is in progress, maybe we can have it in a seperate PR.
3. We had implemented the [discoverability tool tip](https://docs.google.com/document/d/1VUQ7LvAIL8DwBbckUpFCZQJwDSbVEFBBxc2Xt9YWViQ/edit?pli=1#bookmark=id.gt2rdc39umj7), but due to some tests failing, we had removed this. But if it is absolutely necessary, we can prioritise it in this PR.

From the remaining things mentioned, we can take up fixing few things in lesson info modal in this PR and have a separate PR for the rest as you mentioned.

dchen97 commented:
Can we change the text for previously submitted answers to `You answered this question in a previous session.`?

It's fine to not add the discoverability tooltip for now - just wanted to check if it was added

paprajapati9 committed:
review changes and modal changes

paprajapati9 committed:
Merge branch 'oppia:develop' into learner-checkpoint

paprajapati9 commented:
@seanlip Have updated the PR with lesson info modal updates and the change requested by Diana. It's ready to be merged.

paprajapati9 commented:
Attaching the issues left/discussed in this PR, to be worked upon in the follow up PRs: #15466 #15456

seanlip commented:
Thanks @paprajapati9! Merging.

U8NWXD commented:
@paprajapati9 there was just a failure in the frontend tests on develop that I think might be related to this PR: https://github.com/oppia/oppia/runs/6436621000?check_suite_focus=true. Could you please take a look?

paprajapati9 commented:
> @paprajapati9 there was just a failure in the frontend tests on develop that I think might be related to this PR: https://github.com/oppia/oppia/runs/6436621000?check_suite_focus=true. Could you please take a look?

Yes @U8NWXD , I am looking into it

paprajapati9 commented:
@U8NWXD Have raised a follow up PR #15471 that fixes the flakes.
